crates/accelerate/src/inverse_cancellation.rs

```diff
@@ -0,0 +1,193 @@
+                            partitions.push(std::mem::take(&mut chunk));
+                        }
+                    }
+                } else {
+                    panic!("Not an op node");
+                }
+            }
+            for chunk in partitions {
+                if chunk.len() % 2 == 0 {
+                    dag.remove_op_node(chunk[0]);
+                }
+                for node in &chunk[1..] {
+                    dag.remove_op_node(*node);
+                }
+            }
+        }
+    }
+    Ok(())
+}
+fn run_on_inverse_pairs(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op_counts: &IndexMap<String, usize, RandomState>,
+    inverse_gate_names: HashSet<String>,
+    inverse_gates: Vec<[OperationFromPython; 2]>,
+) -> PyResult<()> {
+    if !inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for pair in inverse_gates {
+        let gate_0_name = pair[0].operation.name();
+        let gate_1_name = pair[1].operation.name();
+        if !op_counts.contains_key(gate_0_name) || !op_counts.contains_key(gate_1_name) {
+            continue;
+        }
+        let names: HashSet<String> = pair
+            .iter()
+            .map(|x| x.operation.name().to_string())
+            .collect();
+        let runs: Vec<Vec<NodeIndex>> = dag.collect_runs(names).unwrap().collect();
+        for nodes in runs {
+            let mut i = 0;
+            while i < nodes.len() - 1 {
+                if let NodeType::Operation(inst) = &dag.dag[nodes[i]] {
+                    if let NodeType::Operation(next_inst) = &dag.dag[nodes[i + 1]] {
+                        if inst.qubits == next_inst.qubits
+                            && ((gate_eq(py, inst, &pair[0])? && gate_eq(py, next_inst, &pair[1])?)
+                                || (gate_eq(py, inst, &pair[1])?
+                                    && gate_eq(py, next_inst, &pair[0])?))
+                        {
+                            dag.remove_op_node(nodes[i]);
+                            dag.remove_op_node(nodes[i + 1]);
+                            i += 2;
+                        } else {
+                            i += 1;
+                        }
+                    } else {
+                        panic!("Not an op node")
+                    }
+                } else {
+                    panic!("Not an op node")
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+#[pyfunction]
+pub fn inverse_cancellation(
+    py: Python,
+    dag: &mut DAGCircuit,
+    inverse_gates: Vec<[OperationFromPython; 2]>,
+    self_inverse_gates: Vec<OperationFromPython>,
+    inverse_gate_names: HashSet<String>,
+    self_inverse_gate_names: HashSet<String>,
+) -> PyResult<()> {
+    let op_counts = if !self_inverse_gate_names.is_empty() || !inverse_gate_names.is_empty() {
+        dag.count_ops(py, true)?
+    } else {
+        IndexMap::default()
+    };
```

Comment:
Probably, but I'm not sure it would be necessary. I reworked the logic here to avoid the `default()` call in https://github.com/Qiskit/qiskit/pull/13013/commits/7dee70219033ca759be1b603680d7722765230bb and I think it's a lot more clear.

Reply:
Oh lol yeah, that is much better 😆